tests/test_duplicate_command.rs

```diff
@@ -0,0 +1,110 @@
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::Path;
+
+use crate::common::TestEnvironment;
+
+pub mod common;
+
+fn create_commit(test_env: &TestEnvironment, repo_path: &Path, name: &str, parents: &[&str]) {
+    if parents.is_empty() {
+        test_env.jj_cmd_success(repo_path, &["new", "root", "-m", name]);
+    } else {
+        let mut args = vec!["new", "-m", name];
+        args.extend(parents);
+        test_env.jj_cmd_success(repo_path, &args);
+    }
+    std::fs::write(repo_path.join(name), format!("{name}\n")).unwrap();
+    test_env.jj_cmd_success(repo_path, &["branch", "create", name]);
+}
+
+// https://github.com/martinvonz/jj/issues/1050
+#[test]
+fn test_undo_after_duplicate() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_success(test_env.env_root(), &["init", "repo", "--git"]);
+    let repo_path = test_env.env_root().join("repo");
+
+    create_commit(&test_env, &repo_path, "a", &[]);
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    @ 2443ea76b0b1   a
+    o 000000000000   (no description set)
+    "###);
+
+    let stdout = test_env.jj_cmd_success(&repo_path, &["duplicate", "a"]);
+    insta::assert_snapshot!(stdout, @r###"
+    Created: f5cefcbb65a4 a
+    "###);
+    insta::assert_snapshot!(get_log_output(&test_env, &repo_path), @r###"
+    o f5cefcbb65a4   a
+    | @ 2443ea76b0b1   a
+    |/  
+    o 000000000000   (no description set)
+    "###);
+
+    insta::assert_snapshot!(test_env.jj_cmd_success(&repo_path, &["op", "log"]), @r###"
+    @ 77af4cf52a92 test-username@host.example.com 2001-02-03 04:05:11.000 +07:00 - 2001-02-03 04:05:11.000 +07:00
+    | duplicate commit 2443ea76b0b1c531326908326aab7020abab8e6c
+    | args: jj duplicate a
+    o 0923ccefb5cb test-username@host.example.com 2001-02-03 04:05:09.000 +07:00 - 2001-02-03 04:05:09.000 +07:00
+    | create branch a pointing to commit 2443ea76b0b1c531326908326aab7020abab8e6c
+    | args: jj branch create a
+    o 04db325bf42c test-username@host.example.com 2001-02-03 04:05:09.000 +07:00 - 2001-02-03 04:05:09.000 +07:00
+    | commit working copy
+    o 682c21f25d34 test-username@host.example.com 2001-02-03 04:05:08.000 +07:00 - 2001-02-03 04:05:08.000 +07:00
+    | new empty commit
+    | args: jj new root -m a
+    o a99a3fd5c51e test-username@host.example.com 2001-02-03 04:05:07.000 +07:00 - 2001-02-03 04:05:07.000 +07:00
+    | add workspace 'default'
+    o 56b94dfc38e7 test-username@host.example.com 2001-02-03 04:05:07.000 +07:00 - 2001-02-03 04:05:07.000 +07:00
+      initialize repo
+    "###);
```

Comment:
Do we really need the `jj op log` output to confirm that the undo worked? The regular log output that you have at the end seems enough.

Reply:
No, it was left over from debugging. I'll get rid of it.